File: python/errorta_app/routes/alpha.py
"""F-DIST-01 — local sidecar routes driving the alpha activation/lock UI.

These are the *local* endpoints the Tauri webview calls. They never expose the
license private key (there isn't one on this side) and never leak anything the
check-in service wouldn't already know. Activation is a mutation, so it requires
the Tauri origin header (same guard as the settings routes); status is read-only
and unguarded so the frontend hook can poll it freely.

The actual egress to ``api.errorta.app`` happens inside ``errorta_alpha.client``
— this module only orchestrates it and reports state.
"""
from __future__ import annotations

import threading
import uuid
from pathlib import Path

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

from errorta_alpha import client as alpha_client
from errorta_alpha import config as alpha_config
from errorta_alpha import state as alpha_state
from errorta_alpha import telemetry as alpha_telemetry

router = APIRouter(prefix="/alpha", tags=["alpha"])

# Transient store: a prepared feedback bundle awaiting the tester's explicit
# confirm. Keyed by an opaque id; the path is popped + the file deleted on submit
# so the "show exactly what will be sent" bytes are the bytes that get sent.
# Guarded by a lock: FastAPI runs sync routes on a threadpool, so concurrent
# preview/submit calls must not race the eviction/insert.
_PREPARED_FEEDBACK: dict[str, dict] = {}
_PREPARED_FEEDBACK_CAP = 8
_PREPARED_FEEDBACK_LOCK = threading.Lock()


def _require_tauri_origin(request: Request) -> None:
    origin = request.headers.get("x-errorta-origin", "").lower()
    if origin != "tauri-ui":
        raise HTTPException(status_code=403, detail="tauri origin required")


class ActivateRequest(BaseModel):
    code: str = Field(min_length=1, max_length=64)


@router.get("/status")
def status() -> dict:
    """Current alpha state (drives the activation screen / lock screen / chrome
    gating). Read-only; safe to poll."""
    return alpha_state.current_status().to_dict()


@router.post("/activate")
def activate(body: ActivateRequest, request: Request) -> dict:
    """Redeem an invite code against the check-in service and persist the license.

    On success returns the refreshed status. On a rejected code, surfaces the
    service's error code (``code_not_found`` / ``code_exhausted`` /
    ``code_disabled`` / ``code_expired`` / ``device_code_mismatch``) or
    ``offline`` as a 400 so the UI can render honest copy.
    """
    _require_tauri_origin(request)
    if not alpha_config.gate_enabled():
        raise HTTPException(status_code=409, detail="alpha gate is not enabled in this build")

    result = alpha_client.activate(body.code)
    if not result.ok:
        raise HTTPException(
            status_code=400,
            detail={"error": result.error_code or "activation_failed", "message": result.message},
        )
    return alpha_state.current_status().to_dict()


class TelemetryPut(BaseModel):
    extras_enabled: bool


@router.get("/telemetry")
def get_telemetry() -> dict:
    """Consent state for the Settings → Alpha telemetry panel."""
    return {
        "gate_enabled": alpha_config.gate_enabled(),
        "extras_enabled": alpha_telemetry.extras_enabled(),
    }


@router.put("/telemetry")
def put_telemetry(body: TelemetryPut, request: Request) -> dict:
    """Turn the Tier-2 extras on/off. The disclosed Tier-1 floor is not
    opt-out-able while enrolled (spec §9), so only ``extras_enabled`` is settable."""
    _require_tauri_origin(request)
    alpha_telemetry.set_extras_enabled(body.extras_enabled)
    return {"extras_enabled": alpha_telemetry.extras_enabled()}


@router.get("/telemetry/inspect")
def inspect_telemetry() -> dict:
    """The "see exactly what we send" payload: the pending floor deltas + the
    tail of the extras queue, verbatim. Nothing is ever sent that isn't visible
    here first."""
    return alpha_telemetry.inspector_snapshot()


class FeedbackPreviewRequest(BaseModel):
    kind: str = Field(default="bug", max_length=32)
    message: str = Field(default="", max_length=8000)


@router.post("/feedback/preview")
def feedback_preview(body: FeedbackPreviewRequest, request: Request) -> dict:
    """Build the redacted feedback bundle and return its manifest so the tester
    can review EXACTLY what will be sent before confirming. Reachable regardless
    of lock state — a locked/unactivated tester can still tell us what's wrong."""
    _require_tauri_origin(request)
    if not alpha_config.gate_enabled():
        raise HTTPException(status_code=409, detail="alpha gate is not enabled in this build")
    from errorta_alpha import feedback as alpha_feedback

    log_buffer = getattr(request.app.state, "log_buffer", None)
    result = alpha_feedback.prepare_feedback_bundle(user_note=body.message, log_buffer=log_buffer)

    prepared_id = uuid.uuid4().hex
    with _PREPARED_FEEDBACK_LOCK:
        # Bound the transient store; drop the OLDEST prepared bundle(s) + their
        # files. dicts preserve insertion order, so the first key is the oldest.
        while len(_PREPARED_FEEDBACK) >= _PREPARED_FEEDBACK_CAP:
            oldest = next(iter(_PREPARED_FEEDBACK))
            stale = _PREPARED_FEEDBACK.pop(oldest)
            Path(str(stale.get("path", ""))).unlink(missing_ok=True)
        _PREPARED_FEEDBACK[prepared_id] = {
            "path": str(result["path"]),
            "kind": body.kind,
            "message": body.message,
        }
    return {
        "prepared_id": prepared_id,
        "kind": body.kind,
        "message": body.message,
        "bundle": {
            "sha256": result.get("sha256"),
            "files": result.get("files"),
            "redaction": result.get("redaction_manifest"),
        },
    }


class FeedbackSubmitRequest(BaseModel):
    prepared_id: str = Field(min_length=1, max_length=64)


@router.post("/feedback/submit")
def feedback_submit(body: FeedbackSubmitRequest, request: Request) -> dict:
    """Send a previously-previewed feedback bundle. The prepared bundle is popped
    and its file deleted so the previewed bytes are exactly the sent bytes."""
    _require_tauri_origin(request)
    if not alpha_config.gate_enabled():
        raise HTTPException(status_code=409, detail="alpha gate is not enabled in this build")
    with _PREPARED_FEEDBACK_LOCK:
        prep = _PREPARED_FEEDBACK.pop(body.prepared_id, None)
    if prep is None:
        raise HTTPException(status_code=404, detail="unknown or expired prepared feedback")
    result = alpha_client.send_feedback(
        kind=prep["kind"], message=prep["message"], bundle_path=prep["path"]
    )
    Path(str(prep["path"])).unlink(missing_ok=True)
    if not result.ok:
        raise HTTPException(status_code=502, detail={"error": result.error or "send_failed"})
    return {"ticket_id": result.ticket_id}

File: src/lib/api/alpha.ts
// F-DIST-01 — typed client for the local /alpha/* sidecar routes that drive the
// activation + lock UI. The sidecar owns the license state and the (disclosed)
// egress to api.errorta.app; this client only reads status and submits a code.

import { getJSON, sidecarFetch, SidecarUnreachableError } from "../api";

export type AlphaState = "disabled" | "unactivated" | "active" | "expired" | "revoked";

export interface AlphaStatus {
  gateEnabled: boolean;
  state: AlphaState;
  locked: boolean;
  reason: string | null;
  graceUntil: number | null;
  deviceId: string | null;
  buildEol: boolean;
  buildEolRequired: boolean;
  updateUrl: string | null;
}

interface AlphaStatusWire {
  gate_enabled: boolean;
  state: AlphaState;
  locked: boolean;
  reason: string | null;
  grace_until: number | null;
  device_id: string | null;
  build_eol?: boolean;
  build_eol_required: boolean;
  update_url: string | null;
}

function adapt(w: AlphaStatusWire): AlphaStatus {
  return {
    gateEnabled: w.gate_enabled,
    state: w.state,
    locked: w.locked,
    reason: w.reason,
    graceUntil: w.grace_until,
    deviceId: w.device_id,
    buildEol: w.build_eol ?? false,
    buildEolRequired: w.build_eol_required,
    updateUrl: w.update_url,
  };
}

export async function getAlphaStatus(): Promise<AlphaStatus> {
  return adapt(await getJSON<AlphaStatusWire>("/alpha/status"));
}

/** Thrown when activation is rejected; `code` is the machine-readable reason
 *  from the check-in service (e.g. `code_exhausted`, `offline`). */
export class AlphaActivationError extends Error {
  constructor(
    public readonly code: string,
    message?: string,
  ) {
    super(message || code);
    this.name = "AlphaActivationError";
  }
}

const UI_HEADERS = { "x-errorta-origin": "tauri-ui" };

// Back-off schedule for a transient activation transport failure. The webview
// can race the sidecar during first-launch warm-up / respawn. A rejected fetch
// does not prove whether the sidecar processed the request before the response
// was lost, so this retry relies on /v1/activate being idempotent for the same
// device_id + code (the Worker returns the existing seat without incrementing).
const ACTIVATE_BACKOFF_MS = [300, 700, 1500];
const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

export async function activateAlpha(code: string): Promise<AlphaStatus> {
  // A `SidecarUnreachableError` means no HTTP response reached the webview. The
  // activation protocol is idempotent for the same device + code, so retrying is
  // safe even if an earlier request committed before its response was lost. A
  // real HTTP response (server-side rejection) throws `AlphaActivationError`
  // below and is NOT retried (deterministic). This closes
  // the "first launch → Activate raced the still-booting sidecar → generic
  // failure" gap that a tester can't work around.
  for (let attempt = 0; ; attempt++) {
    let res: Response;
    try {
      res = await sidecarFetch("/alpha/activate", {
        method: "POST",
        headers: { ...UI_HEADERS, "content-type": "application/json" },
        body: JSON.stringify({ code }),
      });
    } catch (err) {
      if (err instanceof SidecarUnreachableError && attempt < ACTIVATE_BACKOFF_MS.length) {
        await sleep(ACTIVATE_BACKOFF_MS[attempt]);
        continue;
      }
      throw err;
    }
    if (res.ok) {
      return adapt((await res.json()) as AlphaStatusWire);
    }
    // FastAPI serializes HTTPException(detail={...}) as {"detail": {error, message}}.
    let errorCode = `http_${res.status}`;
    let message: string | undefined;
    try {
      const body = (await res.json()) as { detail?: { error?: string; message?: string } };
      if (body.detail?.error) errorCode = body.detail.error;
      message = body.detail?.message ?? undefined;
    } catch {
      // non-JSON error body — keep the http_<status> code
    }
    throw new AlphaActivationError(errorCode, message);
  }
}

// --- telemetry (slice 6) -----------------------------------------------------

export interface AlphaTelemetryConsent {
  gateEnabled: boolean;
  extrasEnabled: boolean;
}

export interface AlphaTelemetryEvent {
  event: string;
  name?: string;
  bucket?: string;
  count?: number;
}

export interface AlphaTelemetryInspect {
  extrasEnabled: boolean;
  floor: Record<string, number>;
  queue: AlphaTelemetryEvent[];
  queueLen: number;
}

export async function getTelemetryConsent(): Promise<AlphaTelemetryConsent> {
  const w = await getJSON<{ gate_enabled: boolean; extras_enabled: boolean }>("/alpha/telemetry");
  return { gateEnabled: w.gate_enabled, extrasEnabled: w.extras_enabled };
}

export async function setTelemetryExtras(enabled: boolean): Promise<boolean> {
  const res = await sidecarFetch("/alpha/telemetry", {
    method: "PUT",
    headers: { ...UI_HEADERS, "content-type": "application/json" },
    body: JSON.stringify({ extras_enabled: enabled }),
  });
  if (!res.ok) throw new Error(`PUT /alpha/telemetry failed (${res.status})`);
  const w = (await res.json()) as { extras_enabled: boolean };
  return w.extras_enabled;
}

export async function getTelemetryInspect(): Promise<AlphaTelemetryInspect> {
  const w = await getJSON<{
    extras_enabled: boolean;
    floor: Record<string, number>;
    queue: AlphaTelemetryEvent[];
    queue_len: number;
  }>("/alpha/telemetry/inspect");
  return {
    extrasEnabled: w.extras_enabled,
    floor: w.floor ?? {},
    queue: w.queue ?? [],
    queueLen: w.queue_len ?? 0,
  };
}

// --- feedback (slice 7) ------------------------------------------------------

export type FeedbackKind = "bug" | "suggestion" | "crash";

export interface FeedbackPreview {
  preparedId: string;
  kind: FeedbackKind;
  message: string;
  bundle: {
    sha256: string | null;
    files: string[];
    redaction: Record<string, number>;
  };
}

/** Build the redacted bundle and return its manifest so the tester sees exactly
 *  what will be sent before confirming. */
export async function previewFeedback(
  kind: FeedbackKind,
  message: string,
): Promise<FeedbackPreview> {
  const res = await sidecarFetch("/alpha/feedback/preview", {
    method: "POST",
    headers: { ...UI_HEADERS, "content-type": "application/json" },
    body: JSON.stringify({ kind, message }),
  });
  if (!res.ok) throw new Error(`preview feedback failed (${res.status})`);
  const w = (await res.json()) as {
    prepared_id: string;
    kind: FeedbackKind;
    message: string;
    bundle: { sha256: string | null; files?: string[]; redaction?: Record<string, number> };
  };
  return {
    preparedId: w.prepared_id,
    kind: w.kind,
    message: w.message,
    bundle: {
      sha256: w.bundle.sha256,
      files: w.bundle.files ?? [],
      redaction: w.bundle.redaction ?? {},
    },
  };
}

/** Send a previously-previewed feedback bundle. Returns the ticket id. */
export async function submitFeedback(preparedId: string): Promise<string> {
  const res = await sidecarFetch("/alpha/feedback/submit", {
    method: "POST",
    headers: { ...UI_HEADERS, "content-type": "application/json" },
    body: JSON.stringify({ prepared_id: preparedId }),
  });
  if (!res.ok) throw new Error(`submit feedback failed (${res.status})`);
  const w = (await res.json()) as { ticket_id: string };
  return w.ticket_id;
}
